<a href="https://colab.research.google.com/github/dataqueenpend/Projects-/blob/main/Testing%2C_testing_housepricespynb.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
from sklearn import model_selection

from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from xgboost import XGBRegressor

In [2]:
df_train = pd.read_csv("/content/train_after_fe.csv")
df_test = pd.read_csv("/content/test_after_fe.csv")

In [3]:
df_train.head()

Id  SalePrice  MSSubClass  ...  LotFrontage NaN  MasVnrArea NaN  GarageYrBlt NaN
0   1  12.247694    0.235294  ...              0.0             0.0              0.0
1   2  12.109011    0.000000  ...              0.0             0.0              0.0
2   3  12.317167    0.235294  ...              0.0             0.0              0.0
3   4  11.849398    0.294118  ...              0.0             0.0              0.0
4   5  12.429216    0.235294  ...              0.0             0.0              0.0

[5 rows x 84 columns]

In [4]:
df_test.head()

Id  MSSubClass  ...  BsmtHalfBath NaN  GarageYrBlt NaN
0  1461    0.000000  ...               0.0              0.0
1  1462    0.000000  ...               0.0              0.0
2  1463    0.235294  ...               0.0              0.0
3  1464    0.235294  ...               0.0              0.0
4  1465    0.588235  ...               0.0              0.0

[5 rows x 85 columns]

In [5]:
#y= df_train['SalePrice']

In [6]:
#df_train = df_train.drop('SalePrice', axis=1)

In [7]:
df_train["kfold"] = -1

In [8]:
kf = model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
for fold, (train_indicies, valid_indicies) in enumerate(kf.split(X=df_train)):
    df_train.loc[valid_indicies, "kfold"] = fold

In [9]:
useful_features = [c for c in df_train.columns if c not in ("Id", "SalePrice", "kfold")]

In [10]:
df_test = df_test[useful_features]

In [11]:
final_predictions = []
for fold in range(5):
    xtrain =  df_train[df_train.kfold != fold].reset_index(drop=True)
    xvalid = df_train[df_train.kfold == fold].reset_index(drop=True)
    xtest = df_test.copy()

    ytrain = xtrain.SalePrice
    yvalid = xvalid.SalePrice
    
    xtrain = xtrain[useful_features]
    xvalid = xvalid[useful_features]

    model = XGBRegressor(random_state=fold, n_jobs=4)
    model.fit(xtrain, ytrain)
    preds_valid = model.predict(xvalid)
    test_preds = model.predict(xtest)
    final_predictions.append(test_preds)
    print(fold, mean_squared_error(yvalid, preds_valid, squared=False))

[16:34:33] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
0 0.13205370062118182
[16:34:34] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
1 0.11618442501334689
[16:34:34] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
2 0.15560101668402265
[16:34:35] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
3 0.13592830039405301
[16:34:35] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
4 0.12104948767090469


woah. this is reallly interesting. the model is performing absurdly bad. let's check out with raw data and simple pandas tricks

In [12]:
df_raw = pd.read_csv('/content/train.csv')
df_raw_test = pd.read_csv('/content/test.csv')

In [13]:
df_for_features_eng = df_raw.drop('SalePrice', axis=1)

In [14]:
df_whole = pd.concat([df_for_features_eng, df_raw_test])

In [15]:
df_whole.head()

Id  MSSubClass MSZoning  LotFrontage  ...  MoSold YrSold SaleType SaleCondition
0   1          60       RL         65.0  ...       2   2008       WD        Normal
1   2          20       RL         80.0  ...       5   2007       WD        Normal
2   3          60       RL         68.0  ...       9   2008       WD        Normal
3   4          70       RL         60.0  ...       2   2006       WD       Abnorml
4   5          60       RL         84.0  ...      12   2008       WD        Normal

[5 rows x 80 columns]

In [16]:
df_whole.shape

(2919, 80)

In [17]:
nan_s = df_whole.isnull().sum()
only_nans = nan_s[nan_s > 0].sort_values(ascending=False)
print(only_nans)

PoolQC          2909
MiscFeature     2814
Alley           2721
Fence           2348
FireplaceQu     1420
LotFrontage      486
GarageFinish     159
GarageYrBlt      159
GarageQual       159
GarageCond       159
GarageType       157
BsmtExposure      82
BsmtCond          82
BsmtQual          81
BsmtFinType2      80
BsmtFinType1      79
MasVnrType        24
MasVnrArea        23
MSZoning           4
BsmtFullBath       2
BsmtHalfBath       2
Utilities          2
Functional         2
Exterior2nd        1
Exterior1st        1
SaleType           1
BsmtFinSF1         1
BsmtFinSF2         1
BsmtUnfSF          1
Electrical         1
KitchenQual        1
GarageCars         1
GarageArea         1
TotalBsmtSF        1
dtype: int64


##Cat features, num features selection (nan's)

In [18]:
cat_cols_nans = [col for col in df_whole.columns if (df_whole[col].dtype == 'object') and (df_whole[col].isnull().any())]

In [19]:
df_cols_nans = df_whole[cat_cols_nans]

In [20]:
df_cols_nans.isnull().sum().sort_values(ascending=False)

PoolQC          2909
MiscFeature     2814
Alley           2721
Fence           2348
FireplaceQu     1420
GarageCond       159
GarageQual       159
GarageFinish     159
GarageType       157
BsmtExposure      82
BsmtCond          82
BsmtQual          81
BsmtFinType2      80
BsmtFinType1      79
MasVnrType        24
MSZoning           4
Functional         2
Utilities          2
Exterior1st        1
SaleType           1
Exterior2nd        1
KitchenQual        1
Electrical         1
dtype: int64

In [21]:
len(df_cols_nans.columns)

23

Okay we see in what columns there are nan's values. 
For categorical features we'll replace nan's with zero - meaning no information (which is usually the case if there is no info in categorical feature). 
> For example we have 2909 nan's in Pool columns (on 2919 possible), which it obviously indicates that most of the people don't have a pool. 

> I would also drop totally MiscFeature column which doesn't tell us anything about the house. 

We could also be more mindful about dealing with nan's, but for the quick analysis it will suit us.

In [22]:
df_whole = df_whole.drop('MiscFeature', axis=1)

In [23]:
cat_cols_nans.remove('MiscFeature')

In [24]:
df_whole[cat_cols_nans] = df_whole[cat_cols_nans].fillna('Missing')

In [25]:
#Checking the nans
df_whole[cat_cols_nans].isnull().sum()

MSZoning        0
Alley           0
Utilities       0
Exterior1st     0
Exterior2nd     0
MasVnrType      0
BsmtQual        0
BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinType2    0
Electrical      0
KitchenQual     0
Functional      0
FireplaceQu     0
GarageType      0
GarageFinish    0
GarageQual      0
GarageCond      0
PoolQC          0
Fence           0
SaleType        0
dtype: int64

In [26]:
#Num features selection 
num_cols_nans = [col for col in df_whole.columns if (df_whole[col].dtype != 'object') and (df_whole[col].isnull().any())]

In [27]:
df_cols_num_nans = df_whole[num_cols_nans]

In [28]:
df_cols_num_nans.isnull().sum().sort_values(ascending=False)

LotFrontage     486
GarageYrBlt     159
MasVnrArea       23
BsmtHalfBath      2
BsmtFullBath      2
GarageArea        1
GarageCars        1
TotalBsmtSF       1
BsmtUnfSF         1
BsmtFinSF2        1
BsmtFinSF1        1
dtype: int64

In [29]:
len(num_cols_nans)

11

Also it seems like some of the numerical columns are in truth categorical (garage cars, half bath, full bath). Also there are some Yr columns in here... But for the purpose of our quick analysis we'll just replace nans with mean's of the individual columns.

In [30]:
for col in num_cols_nans:
  median = df_whole[col].median()
  df_whole[col].fillna(median, inplace=True)

In [31]:
#Checking num cols for nans
df_whole[num_cols_nans].isnull().sum()

LotFrontage     0
MasVnrArea      0
BsmtFinSF1      0
BsmtFinSF2      0
BsmtUnfSF       0
TotalBsmtSF     0
BsmtFullBath    0
BsmtHalfBath    0
GarageYrBlt     0
GarageCars      0
GarageArea      0
dtype: int64

In [32]:
#Last check for the whole df
df_whole.isnull().sum()

Id               0
MSSubClass       0
MSZoning         0
LotFrontage      0
LotArea          0
                ..
MiscVal          0
MoSold           0
YrSold           0
SaleType         0
SaleCondition    0
Length: 79, dtype: int64

No nan's. 

Next?


##Dummy variables

As everyone knows our machine don't like not number values. So in order to make them like our code, we need to impute the categorical features from text values to numerical. 

There are multiple ways to do that. We'll just use simple get_dummies from Pandas. But **beware**! Do you know about the trap of pandas dummies? If not go to towards data science [article]('https://towardsdatascience.com/beware-of-the-dummy-variable-trap-in-pandas-727e8e6b8bde'). 

In [41]:
df_whole = pd.get_dummies(df_whole)

In [43]:
#Let's take a look at our dataset
df_whole.head()

Id  MSSubClass  ...  SaleCondition_Normal  SaleCondition_Partial
0   1          60  ...                     1                      0
1   2          20  ...                     1                      0
2   3          60  ...                     1                      0
3   4          70  ...                     0                      0
4   5          60  ...                     1                      0

[5 rows x 307 columns]

In [45]:
df_whole.shape

(2919, 307)

Woah that's a lot of columns!

> Of course we could reduce this number by deleting rare labels first. Maybe we'll do it later.

Now I want to just test the data for the performance. Without scaling, without dealing with outliers etc.  

##Performance check

In [46]:
y = df_raw['SalePrice']

In [57]:
df_train = df_whole.iloc[:df_raw.shape[0]]
df_test = df_whole.iloc[df_raw.shape[0]:]

In [66]:
df_train['SalePrice'] = df_raw['SalePrice']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [67]:
df_train.head()

Id  MSSubClass  ...  SaleCondition_Partial  SalePrice
0   1          60  ...                      0     208500
1   2          20  ...                      0     181500
2   3          60  ...                      0     223500
3   4          70  ...                      0     140000
4   5          60  ...                      0     250000

[5 rows x 308 columns]

In [68]:
df_train["kfold"] = -1

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [69]:
kf = model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
for fold, (train_indicies, valid_indicies) in enumerate(kf.split(X=df_train)):
    df_train.loc[valid_indicies, "kfold"] = fold

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [78]:
useful_features = [c for c in df_train.columns if c not in ("Id", "SalePrice", "kfold")]
#object_cols = [col for col in useful_features if 'cat' in col]
df_test = df_test[useful_features]

In [79]:

final_predictions = []
for fold in range(5):
    xtrain =  df_train[df_train.kfold != fold].reset_index(drop=True)
    xvalid = df_train[df_train.kfold == fold].reset_index(drop=True)
    xtest = df_test.copy()

    ytrain = xtrain.SalePrice
    yvalid = xvalid.SalePrice
    
    xtrain = xtrain[useful_features]
    xvalid = xvalid[useful_features]
    
    
    model = XGBRegressor(random_state=fold, n_jobs=4)
    model.fit(xtrain, ytrain)
    preds_valid = model.predict(xvalid)
    test_preds = model.predict(xtest)
    final_predictions.append(test_preds)
    print(fold, mean_squared_error(yvalid, preds_valid, squared=False))

[15:28:57] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
0 29957.172528108797
[15:28:58] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
1 28517.005613200807
[15:28:59] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
2 47361.566142760144
[15:29:00] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
3 26607.78125490112
[15:29:00] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
4 22166.15985201308


It seems like it performed better? Let's check out the predictions. 

In [80]:
preds = np.mean(np.column_stack(final_predictions), axis=1)

In [81]:
sample_submission = pd.read_csv("/content/sample_submission.csv")

In [82]:
sample_submission.target = preds
sample_submission.to_csv("submission_kfolds_without_transformations.csv", index=False)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  """Entry point for launching an IPython kernel.


Yep! Hooray! Finally some improvement :D

Next I'll be checking out how the model will perform with other types of data transformation. First let's remove rare labels from the data. 

##Removing rare labels

In [33]:
#Changing rare labels to 'Rare Label' string
cat_cols = [col for col in df_whole.columns if df_whole[col].dtype == 'object']

for col in cat_cols:
  temp = df_whole.groupby(col)['Id'].count()/len(df_whole)
  temp_df = temp[temp >0.01].index
  df_whole[col] = np.where(df_whole[col].isin(temp_df), df_whole[col], 'Rare label')

In [35]:
df_whole.head(50)

Id  MSSubClass    MSZoning  ...  YrSold  SaleType SaleCondition
0    1          60          RL  ...    2008        WD        Normal
1    2          20          RL  ...    2007        WD        Normal
2    3          60          RL  ...    2008        WD        Normal
3    4          70          RL  ...    2006        WD       Abnorml
4    5          60          RL  ...    2008        WD        Normal
5    6          50          RL  ...    2009        WD        Normal
6    7          20          RL  ...    2007        WD        Normal
7    8          60          RL  ...    2009        WD        Normal
8    9          50          RM  ...    2008        WD       Abnorml
9   10         190          RL  ...    2008        WD        Normal
10  11          20          RL  ...    2008        WD        Normal
11  12          60          RL  ...    2006       New       Partial
12  13          20          RL  ...    2008        WD        Normal
13  14          20          RL  ...    2007       New       Partial
14  15          20          RL  ...    2008        WD        Normal
15  16          45          RM  ...    2007        WD        Normal
16  17          20          RL  ...    2010        WD        Normal
17  18          90          RL  ...    2006        WD        Normal
18  19          20          RL  ...    2008        WD        Normal
19  20          20          RL  ...    2009       COD       Abnorml
20  21          60          RL  ...    2006       New       Partial
21  22          45          RM  ...    2007        WD        Normal
22  23          20          RL  ...    2008        WD        Normal
23  24         120          RM  ...    2007        WD        Normal
24  25          20          RL  ...    2010        WD        Normal
25  26          20          RL  ...    2009        WD        Normal
26  27          20          RL  ...    2010        WD        Normal
27  28          20          RL  ...    2010        WD        Normal
28  29          20          RL  ...    2006        WD        Normal
29  30          30          RM  ...    2008        WD        Normal
30  31          70  Rare label  ...    2008        WD        Normal
31  32          20          RL  ...    2008        WD        Normal
32  33          20          RL  ...    2008        WD        Normal
33  34          20          RL  ...    2010        WD        Normal
34  35         120          RL  ...    2007        WD        Normal
35  36          60          RL  ...    2006        WD        Normal
36  37          20          RL  ...    2009        WD        Normal
37  38          20          RL  ...    2009        WD        Normal
38  39          20          RL  ...    2010        WD       Abnorml
39  40          90          RL  ...    2008        WD    Rare label
40  41          20          RL  ...    2006        WD       Abnorml
41  42          20          RL  ...    2007        WD        Normal
42  43          85          RL  ...    2007        WD        Normal
43  44          20          RL  ...    2008        WD        Normal
44  45          20          RL  ...    2006        WD        Normal
45  46         120          RL  ...    2010        WD        Normal
46  47          50          RL  ...    2009        WD       Abnorml
47  48          20          FV  ...    2007        WD        Normal
48  49         190          RM  ...    2009       New       Partial
49  50          20          RL  ...    2007        WD        Normal

[50 rows x 79 columns]

In [43]:
#Creating dummies
df_whole = pd.get_dummies(df_whole, drop_first=True)

In [53]:
df_whole.head()

Id  MSSubClass  ...  SaleCondition_Partial  SaleCondition_Rare label
0   1          60  ...                      0                         0
1   2          20  ...                      0                         0
2   3          60  ...                      0                         0
3   4          70  ...                      0                         0
4   5          60  ...                      0                         0

[5 rows x 202 columns]

##Performance check after removing labels and getting dummies

In [54]:
df_train = df_whole.iloc[:df_raw.shape[0]]
df_test = df_whole.iloc[df_raw.shape[0]:]

In [55]:
df_train.head()

Id  MSSubClass  ...  SaleCondition_Partial  SaleCondition_Rare label
0   1          60  ...                      0                         0
1   2          20  ...                      0                         0
2   3          60  ...                      0                         0
3   4          70  ...                      0                         0
4   5          60  ...                      0                         0

[5 rows x 202 columns]

In [56]:
df_test.head()

Id  MSSubClass  ...  SaleCondition_Partial  SaleCondition_Rare label
0  1461          20  ...                      0                         0
1  1462          20  ...                      0                         0
2  1463          60  ...                      0                         0
3  1464          60  ...                      0                         0
4  1465         120  ...                      0                         0

[5 rows x 202 columns]

In [57]:
df_train['SalePrice'] = df_raw['SalePrice']

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [58]:
df_train["kfold"] = -1

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [59]:
kf = model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
for fold, (train_indicies, valid_indicies) in enumerate(kf.split(X=df_train)):
    df_train.loc[valid_indicies, "kfold"] = fold

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [60]:
useful_features = [c for c in df_train.columns if c not in ("Id", "SalePrice", "kfold")]
object_cols = [col for col in df_train.columns if (df_train[col].dtype == 'object')]
df_test = df_test[useful_features]

In [61]:

final_predictions = []
for fold in range(5):
    xtrain =  df_train[df_train.kfold != fold].reset_index(drop=True)
    xvalid = df_train[df_train.kfold == fold].reset_index(drop=True)
    xtest = df_test.copy()

    ytrain = xtrain.SalePrice
    yvalid = xvalid.SalePrice
    
    xtrain = xtrain[useful_features]
    xvalid = xvalid[useful_features]
 
    model = XGBRegressor(random_state=fold, n_jobs=4)
    model.fit(xtrain, ytrain)
    preds_valid = model.predict(xvalid)
    test_preds = model.predict(xtest)
    final_predictions.append(test_preds)
    print(fold, mean_squared_error(yvalid, preds_valid, squared=False))

[16:51:27] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
0 30949.34052908671
[16:51:27] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
1 28475.729167266294
[16:51:28] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
2 47856.803477161404
[16:51:28] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
3 27672.347236827296
[16:51:29] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
4 21447.753952428262


In [62]:
preds = np.mean(np.column_stack(final_predictions), axis=1)

In [63]:
sample_submission = pd.read_csv('/content/sample_submission.csv')

In [65]:
sample_submission.to_csv('rare_labels_dummies_xgb.csv', index=False)

Interestingly after removing rare labels model achieved not so much better (only in one case better) results. 